Sources/PackagePlugin/Plugin.swift

```diff
@@ -72,15 +95,35 @@ extension Plugin {
             }
             
             // Ask the plugin to create build commands for the input target.
-            commands = try plugin.createBuildCommands(context: context, target: target)
+            generatedCommands = try plugin.createBuildCommands(context: context, target: target)
+            
+        case .performCommand(let targets, let arguments):
+            // Check that the plugin implements the appropriate protocol for its
+            // declared capability.
+            guard let plugin = plugin as? CommandPlugin else {
+                throw PluginDeserializationError.malformedInputJSON("Plugin declared with `command` capability but doesn't conform to `CommandPlugin` protocol")
+            }
+            
+            // Invoke the plugin.
+            try plugin.performCommand(context: context, targets: targets, arguments: arguments)
+
+            // For command plugin there are currently no return commands (any
+            // commands invoked by the plugin are invoked directly).
+            generatedCommands = []
         }
         
-        // Construct the output structure to send to SwiftPM.
-        let output = try PluginOutput(commands: commands, diagnostics: Diagnostics.emittedDiagnostics)
+        // Construct the output structure to send back to SwiftPM.
+        let output: PluginOutput
+        do {
+            output = try PluginOutput(commands: generatedCommands, diagnostics: Diagnostics.emittedDiagnostics)
+        } catch {
+            internalError("Couldn’t encode output JSON: \(error).")
+        }
 
         // On stdout, write a zero byte followed by the JSON data — this is what libSwiftPM expects to see. Anything before the last zero byte is treated as freeform output from the plugin (such as debug output from `print` statements). Since `FileHandle.write()` doesn't obey buffering we first have to flush any existing output.
-        fputc(0, stdout)
-        fwrite([UInt8](output.outputData), 1, output.outputData.count, stdout)
+        if fwrite([UInt8](output.outputData), 1, output.outputData.count, jsonOut) != output.outputData.count {
```

Comment:
This one we could probably switch to NSFileHandle.  The main problem with that is that we'd need to raise the macOS minimum deployment target to do it correctly through functions like `write(contentsOf:)`.  Prior to macOS 10.15.4, the NSFileManager APIs could raise ObjC exception for errors instead of throwing catchable Swift errors.  So until we can use those APIs, `fwrite()` seems reasonable.

Reply:
Raising to 10.15.4 sounds fine, since we're already on 10.15 anyway.

Reply:
Can we even express that today though in the package manifest?  Seems to me we'd need to go to 11.

Reply:
I've switched to using the older FileHandle calls in the next iteration of this code, minimizing use of stdio, and then we can switch to the safer ones when we set macOS 11 as minimum deployment target.

Reply:
Ok, it turns out that we can express that using `.macOS("10.15.4")` in the package manifest.  @neonichu @tomerd is that reasonable to include in this PR?  I don't expect many people would be adversely affected, since that was a pretty early 10.15 update.

Reply:
Update: This was merged in a separate PR, so there is no change for that in this PR and I'm using the more modern `FileHandle` APIs.